### PR TITLE
Localizations

### PR DESCRIPTION
**Pull request by jenncoop**
**Summary**

Improvements to localizations

This PR fixes/implements the following **bugs/features**

* [x] Updates truffle to `5.0.0-beta.0`
* [x] `.gitignore` all the things
* [x] Incorporates the `ethereum-localized-messaging` npm lib 
* [x] Fixes broken specs
* [x] Adds new tests related to localization usage

Explain the **motivation** for making this change. What existing problem does the pull request solve?

Abstracts generalized localization logic into lib.

**Test plan (required)**

Followed the same style as the other specs in the test folder. Added new specs for FissionLocalization. Fixed specs that were broken from updating Truffle to beta.

`truffle test`

  Contract: Phone
    #incoming
      on contact list
        ✓ starts the call
        ✓ returns a friendly message
      not on contact list
0xa4
Your message has been recorded
        ✓ goes to answering machine
        ✓ includes the asnwering machine message
    #outgoing
      not on contact list
        ✓ starts the call
        ✓ returns a friendly message
      on contact list
        ✓ starts the call
        ✓ returns a friendly message

  Contract: Status
    #categoryOf
      ✓ retuns the upper nibble (39ms)
      ✓ retuns nibbles above 9
    #reasonOf
      ✓ retuns the upper nibble
      ✓ retuns nibbles above 9
    #isFailure
      lower nibble is 0x00
        ✓ is true
        ✓ is insensitive to the upper nibble
      lower nibble is not 0x00
        ✓ is false
    #isOk
      lower nibble is 0x01
        ✓ is true
        ✓ is insensitive to the upper nibble
      lower nibble is not 0x01
        ✓ is false
    #requireOk/2
      lower nibble is 0x01
        ✓ does not throw (82ms)
      lower nibble is not 0x01
        ✓ reverts with message (65ms)

  Contract: Validators
    AgeValidator
      valid age
        ✓ is okay
        ✓ has the `App` category
      invalid age
        ✓ is not okay
        ✓ has the `App` category
    CombinedValidator
      valid
        ✓ has the `App` category
        ✓ is okay
        ✓ has the insurance validator address in the trace
      invalid
        ✓ has the correct category
        ✓ is not okay
        ✓ has the financial validator address in the trace

  Contract: FissionLocalization
    #log
      ✓ emits a FissionCode event (110ms)
      ✓ emits the correct message (89ms)

**Code formatting**

`npm run test:lint`

This passes quietly

### CODE REVIEW

**Comment by expede** [comment]
> Code formatting
> WIP

@jenncoop friendly FYI: Solidity convention is 4 spaces for indentation. It also may make it easier to look at diffs

File: contracts/localization/FissionLocalization.sol [hunk 1]
```diff
@@ -1,6 +1,6 @@
 pragma solidity ^0.4.24;
 
-import "../../node_modules/ethereum-localized-messaging/contracts/Localization.sol";
+import "/ethereum-localized-messaging/contracts/Localization.sol";
```
Comment (expede) [inline contracts/localization/FissionLocalization.sol:3]: 👍

File: package.json [hunk 2]
```diff
@@ -3,9 +3,13 @@
...
-    "ethereum", "solidity",
-    "status", "code",
-    "eip", "erc", "1066"
+    "ethereum",
+    "solidity",
+    "status",
+    "code",
+    "eip",
+    "erc",
+    "1066"
```
Comment (expede) [inline package.json]: I guess this should also list `erc-1444` in here now, too

Comment (expede) [reply package.json]: Actually, maybe `erc` `1066` should become `erc-1066` as well

File: contracts/Status.sol [hunk 3]
```diff
@@ -1,91 +1,103 @@
...
+    x08,
+    x09,
 
-    enum Reason {
-        Failure,
-        Success,
-        Acceptance,
-        Before,
-        ActionRequired,
+    AppReason,
```
Comment (expede) [inline contracts/Status.sol]: @jenncoop this entire file has 2-space indentation

**Review by expede — CHANGES_REQUESTED** [review]

File: contracts/localization/StatusCodeLocalization.sol [hunk 4]
```diff
@@ -1,15 +0,0 @@
...
-    event StatusCode(byte indexed code, string indexed message);
-
-    function add(byte _code, string _msg) internal {
-        messages[_code] = _msg;
-    }
-
-    function log(byte _code) external {
-        emit StatusCode(_code, messages[_code]);
-    }
-}
```
Comment (expede) [inline contracts/localization/StatusCodeLocalization.sol:15]: 👍

File: contracts/Status.sol (hunk 3 above)
Comment (jenncoop) [reply contracts/Status.sol]: will fix 👍

File: contracts/Status.sol [hunk 5]
```diff
@@ -1,91 +1,103 @@
...
-        return uint(_status >> 4);
-    }
+  // Localization
 
-    function reasonOf(byte _status) public pure returns (uint reason) {
-        return uint(_status & hex"0F");
-    }
+  function localizeBy(byte _status, LocalizationPreferences _prefs) view public returns (bool found, string _msg) {
+    return _prefs.get(_status);
+  }
```
Comment (expede) [inline contracts/Status.sol]: 👍

File: contracts/Status.sol [hunk 6]
```diff
@@ -85,7 +96,8 @@ library Status {
         require(isOk(_status));
     }
 
-    function requireOk(byte _status, string message) public pure {
-        require(isOk(_status), message);
+    function requireOk(byte _status, LocalizationPreferences _prefs) public view {
+        (bool found,) = localizeBy(_status, _prefs);
+        require(found);
     }
```
Comment (expede) [inline contracts/Status.sol:106]: Can we make this overloaded?

```solidity
function requireOk(byte _status, string message) public pure {
    require(isOk(_status), message);
}

function requireOk(byte _status, LocalizationPreferences _prefs) public view {
    (bool _, string message) = localizeBy(_status, _prefs);
    requireOk(_status, message);
}
```

File: contracts/Status.sol [hunk 7]
```diff
@@ -85,7 +96,8 @@ library Status {
         require(isOk(_status));
     }
 
-    function requireOk(byte _status, string message) public pure {
-        require(isOk(_status), message);
+    function requireOk(byte _status, LocalizationPreferences _prefs) public view {
+        (bool found,) = localizeBy(_status, _prefs);
+        require(found);
```
Comment (expede) [inline contracts/Status.sol]: Needs to actually check if `isOk`. The overlaoded comment below includes the expected behaviour

Comment (expede) [reply contracts/Status.sol]: Optional: if you could add a `_` rather than just omitting it. Makes it easier to read IMO.

**Review by expede — CHANGES_REQUESTED** [review]
Getting really close! Just a handful of small comments.

File: contracts/Status.sol (hunk 7 above)
Comment (jenncoop) [reply contracts/Status.sol]: I believe the `_` doesn't compile - I agree it's nicer

File: contracts/Status.sol (hunk 6 above)
Comment (expede) [reply contracts/Status.sol:106]: Oops. Updated

File: contracts/Status.sol (hunk 7 above)
Comment (expede) [reply contracts/Status.sol]: It does compile, but you need to include the type

Comment (expede) [reply contracts/Status.sol]: ie: 

```solidity
(bool _, string storage message)  = ...
```

Comment (expede) [reply contracts/Status.sol]: _because Solidity_ 🤷‍♀️

**Review by expede — APPROVED** [review]
high highest of fives explosion

**Pull request merged**